api/src/main/java/ca/bc/gov/educ/api/gradstudent/service/GradStudentPaginationService.java

```diff
@@ -0,0 +1,153 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.security.InvalidParameterException;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class GradStudentPaginationService {
+  @Getter
+  private final GradStudentPaginationFilterSpecs gradStudentPaginationFilterSpecs;
+
+  private final GradStudentPaginationRepository gradStudentPaginationRepository;
+
+  private final Executor paginatedQueryExecutor = new EnhancedQueueExecutor.Builder()
+    .setThreadFactory(new ThreadFactoryBuilder().setNameFormat("async-pagination-query-executor-%d").build())
+    .setCorePoolSize(2).setMaximumPoolSize(10).setKeepAliveTime(Duration.ofSeconds(60)).build();
+
+  @Transactional(propagation = Propagation.SUPPORTS)
+  public CompletableFuture<Page<GraduationStudentRecordPaginationEntity>> findAll(Specification<GraduationStudentRecordPaginationEntity> studentSpecs, final Integer pageNumber, final Integer pageSize, final List<Sort.Order> sorts) {
+    log.trace("In find all query: {}", studentSpecs);
+    return CompletableFuture.supplyAsync(() -> {
+      Pageable paging = PageRequest.of(pageNumber, pageSize, Sort.by(sorts));
+      try {
+        log.trace("Running paginated query: {}", studentSpecs);
+        var results = this.gradStudentPaginationRepository.findAll(studentSpecs, paging);
+        log.trace("Paginated query returned with results: {}", results);
+        return results;
+      } catch (final Throwable ex) {
+        log.error("Failure querying for paginated SDC school students: {}", ex.getMessage());
+        throw new CompletionException(ex);
+      }
+    }, paginatedQueryExecutor);
+
+  }
+
+  public Specification<GraduationStudentRecordPaginationEntity> setSpecificationAndSortCriteria(String sortCriteriaJson, String searchCriteriaListJson, ObjectMapper objectMapper, List<Sort.Order> sorts) {
+    Specification<GraduationStudentRecordPaginationEntity> schoolSpecs = null;
+    try {
+      RequestUtil.getSortCriteria(sortCriteriaJson, objectMapper, sorts);
+      if (StringUtils.isNotBlank(searchCriteriaListJson)) {
+        List<Search> searches = objectMapper.readValue(searchCriteriaListJson, new TypeReference<>() {
+        });
+        int i = 0;
+        for (var search : searches) {
+          schoolSpecs = getSpecifications(schoolSpecs, i, search, gradStudentPaginationFilterSpecs);
+          i++;
+        }
+      }
+    } catch (JsonProcessingException e) {
+      throw new GradStudentAPIRuntimeException(e.getMessage());
+    }
+    return schoolSpecs;
+  }
+
+  public <T> Specification<T> getSpecifications(Specification<T> specs, int i, Search search, BaseFilterSpecs<T> filterSpecs) {
```

Comment:
Move to a parent class or create a utility class. This uses generics so can be used in other search services (also used in ReportGradStudentSearchService with generics)